cctp_primitives/src/commitment_tree/hashers.rs

```diff
@@ -1,152 +1,233 @@
+    let mut fes = Vec::new();
+
+    // Read constant (if present) as FieldElement and add it to fes
+    if constant.is_some() {
+        fes.push(FieldElement::read(&constant.unwrap()[..])?)
     }
-    if csw_verification_key.is_some(){
-        bytes.extend(&csw_verification_key.unwrap().to_vec());
+
+    // Compute linear hash of custom fields (if present) and add the digest to fes
+    if custom_fields.is_some() {
+        let custom_fes = custom_fields
+            .unwrap()
+            .iter()
+            .map(|custom_field_bytes| FieldElement::read(&custom_field_bytes[..]))
+            .collect::<Result<Vec<_>, _>>()?;
+        fes.push(hash_vec(custom_fes))
     }
-    bytes.extend(&tx_hash.to_vec());
-    bytes.write_u32::<BigEndian>(out_idx)?;
 
-    hash_bytes(&bytes)
-}
+    // Add cert_sysdata_hash
+    fes.push(cert_sysdata_hash);
 
-// Computes FieldElement-based hash on the given Ceased Sidechain Withdrawal data
-pub fn hash_csw(amount: i64,
-                nullifier: &[u8],
-                pk_hash: &[u8],
-                active_cert_data_hash: &[u8])
-    -> Result<FieldElement, Error> {
-    let mut bytes = Vec::<u8>::new();
-
-    bytes.write_i64::<BigEndian>(amount)?;
-    bytes.extend(&nullifier.to_vec());
-    bytes.extend(&pk_hash.to_vec());
-    bytes.extend(&active_cert_data_hash.to_vec());
-
-    hash_bytes(&bytes)
+    // Compute final hash
+    Ok(hash_vec(fes))
 }
 
-// Converts list of BTs to byte-array
-fn bt_list_to_bytes(bt_list: &[(i64,[u8; 20])]) -> Result<Vec<u8>, Error>{
-    let mut bytes = Vec::<u8>::new();
-    for bt in bt_list {
-        bytes.write_i64::<BigEndian>(bt.0)?;
-        bytes.extend(bt.1.to_vec())
-    }
-    Ok(bytes)
+// Computes FieldElement-based hash on the given Certificate data
+//TODO: Maybe we will put additional data ? I hope not :)
+pub fn hash_cert(
+    constant: Option<&[u8; FIELD_SIZE]>,
+    epoch_number: u32,
+    quality: u64,
+    bt_list: &[(u64,[u8; 20])],
+    custom_fields: Option<&[[u8; FIELD_SIZE]]>, //aka proof_data - includes custom_field_elements and bit_vectors merkle roots
+    end_cumulative_sc_tx_commitment_tree_root: &[u8; FIELD_SIZE],
+    btr_fee: u64,
+    ft_min_fee: u64
+) -> Result<FieldElement, Error>
+{
+    get_cert_data_hash(
+        constant, epoch_number, quality, bt_list, custom_fields,
+        end_cumulative_sc_tx_commitment_tree_root, btr_fee, ft_min_fee
+    )
 }
 
-// Computes FieldElement-based hash on the given byte-array
-pub fn hash_bytes(bytes: &[u8]) -> Result<FieldElement, Error> {
-    Ok(hash_vec(&bytes_to_field_elements(bytes)?))
+// Computes FieldElement-based hash on the given Sidechain Creation Transaction data
+pub fn hash_scc(
+    amount: u64,
+    pub_key: &[u8; 32],
+    tx_hash: &[u8; 32],
+    out_idx: u32,
+    withdrawal_epoch_length: u32,
+    cert_proving_system: u8,
+    csw_proving_system: u8,
+    mc_btr_request_data_length: u8,
+    custom_field_elements_configs: &[u8],
```

Comment:
`u8` or `u32` ? Currently `u8` is enough for the tweedle, but what if we decide to use another curve with a size of FE > 256?

Reply:
@alsala in MC and SC code we need to use 1 byte types for `custom_field_elements_configs` and `mc_btr_request_data_length` instead of `int32` to be consistent with CCTP.